x509util/extensions.go

```diff
@@ -500,3 +533,170 @@ func (s *SerialNumber) UnmarshalJSON(data []byte) error {
 	}
 	return nil
 }
+
+// createSubjectAltNameExtension will construct an Extension containing all
+// SubjectAlternativeNames held in a Certificate. It implements more types than
+// the golang x509 library, so it is used whenever OtherName or RegisteredID
+// type SANs are present in the certificate.
+// See also https://datatracker.ietf.org/doc/html/rfc5280.html#section-4.2.1.6
+// TODO: X400Address, DirectoryName, and EDIPartyName types are defined in RFC5280
+//       but are currently unimplemented
+func createSubjectAltNameExtension(c *Certificate) (*Extension, error) {
+	// golang x509 lib does not support all SAN types, to support other types (e.g. otherName, registeredID, etc.)
+	// we need to generate the extension manually
+
+	var rawValues []interface{}
+	allSANs := make([]SubjectAlternativeName, len(c.SANs))
+
+	// First copy in all our known types (DNS, Email, URI, IPs) from the Certificate object into a complete list
+	copy(allSANs, c.SANs)
+
+	for _, dnsName := range c.DNSNames {
+		allSANs = append(allSANs, SubjectAlternativeName{
+			Type:  DNSType,
+			Value: dnsName,
+		})
+	}
+
+	for _, emailAddress := range c.EmailAddresses {
+		allSANs = append(allSANs, SubjectAlternativeName{
+			Type:  EmailType,
+			Value: emailAddress,
+		})
+	}
+
+	for _, uri := range c.URIs {
+		allSANs = append(allSANs, SubjectAlternativeName{
+			Type:  URIType,
+			Value: uri.String(),
+		})
+	}
+
+	for _, ip := range c.IPAddresses {
+		allSANs = append(allSANs, SubjectAlternativeName{
+			Type:  IPType,
+			Value: ip.String(),
+		})
+	}
+
+	// Now iterate over all the SANs and construct the raw ASN1 encoded values and place them in rawValues
+	for _, san := range allSANs {
+
+		switch san.Type {
+
+		case EmailType:
+			rawValues = append(rawValues, asn1.RawValue{Tag: nameTypeEmail, Class: 2, Bytes: []byte(san.Value)})
+		case DNSType:
+			rawValues = append(rawValues, asn1.RawValue{Tag: nameTypeDNS, Class: 2, Bytes: []byte(san.Value)})
+		case X400AddressType:
+			fallthrough
+		case DirectoryNameType:
+			fallthrough
+		case EDIPartyNameType:
+			return nil, fmt.Errorf("unimplemented SAN type %s", san.Type)
+		case URIType:
+			rawValues = append(rawValues, asn1.RawValue{Tag: nameTypeURI, Class: 2, Bytes: []byte(san.Value)})
+		case IPType:
+			if rawIP := net.ParseIP(san.Value); rawIP != nil {
+				ip := rawIP.To4()
+				if ip == nil {
+					ip = rawIP
+				}
+				rawValues = append(rawValues, asn1.RawValue{Tag: nameTypeIP, Class: 2, Bytes: ip})
+			}
+		case RegisteredIDType:
+			if san.Value != "" {
+				oid, err := parseObjectIdentifier(san.Value)
+				if err == nil {
+					rawBytes, _ := asn1.MarshalWithParams(oid, fmt.Sprintf("tag:%d", nameTypeRegisteredID))
+					rawValues = append(rawValues, asn1.RawValue{FullBytes: rawBytes})
+				}
+			}
+		default:
```

Comment:
We have also the AutoType and "" that is equivalent to AutoType, see `SubjectAlternativeName.Set()`.